topdown/eval.go

```diff
@@ -31,6 +31,7 @@ type eval struct {
 	parent        *eval
 	cancel        Cancel
 	query         ast.Body
+	currExpr      *ast.Expr
```

Comment:
I'd rather not have the state for the current expression duplicated like this. The evaluator keeps the index of the current expression on the stack and passes it around as necessary.

I can imagine two solutions here that avoid duplicating state like this...

1a. Update the eval struct to include the index. When stepping to the next expression, make a copy of the eval struct and bump the index.

1b. Same as 1a but avoid copying eval struct--instead increment/decrement the index around each step. This might be cheaper than 1a. Would need to benchmark and find out.

2. Pass the index down to where it's needed either via the evalFunc and evalTree/evalVirtual structs or as params.

I think 1a or 1b are probably the most useful as anything that has access to the eval struct can lookup the currently executing expression without needing to be modified to accept the index.